polkadot/runtime/parachains/src/assigner_on_demand/mod.rs

```diff
@@ -110,26 +110,41 @@ pub enum SpotTrafficCalculationErr {
 }
 
 /// Assignments as provided by the on-demand `AssignmentProvider`.
-#[derive(RuntimeDebug, Encode, Decode, TypeInfo)]
+#[derive(RuntimeDebug, Encode, Decode, TypeInfo, PartialEq, Clone)]
 pub struct OnDemandAssignment {
 	/// The assigned para id.
 	para_id: ParaId,
 	/// The core index the para got assigned to.
 	core_index: CoreIndex,
 }
 
+#[cfg(test)]
+impl OnDemandAssignment {
+	pub fn new(para_id: ParaId, core_index: CoreIndex) -> Self {
+		Self { para_id, core_index }
+	}
+}
+
 impl Assignment for OnDemandAssignment {
 	fn para_id(&self) -> ParaId {
 		self.para_id
 	}
 }
 
 /// Internal representation of an order after it has been enqueued already.
-#[derive(Encode, Decode, TypeInfo)]
+#[cfg(not(test))]
+#[derive(Encode, Decode, TypeInfo, Debug, PartialEq)]
 pub(super) struct EnqueuedOrder {
 	pub para_id: ParaId,
 }
 
+// Looser encapsulation for tests
```

Comment:
Can we re-organize tests, so this is not necessary? I see that tests in the scheduler are accessing this, but this seems a bit off from a high level perspective. Scheduler tests should not bother with internal state of the on-demand assigner. It is likely that these checks should be replaces with tests in the on-demand assigner itself.

Reply:
I think we can definitely do this. It seems there are two ways the scheduler tests are breaking encapsulation of the assigner.
1. Calling interface that should be private like `add_on_demand_order` rather than using the more complex public interface, `place_order_keep_alive`. This requires exposing the otherwise private function and the `EnqueuedOrder` parameter type.
2. Constructor for `Assignment`, needed to produce expected `ParasEntry` results. 

For 1, I'd like to use `place_order_keep_alive` rather than `add_on_demand_order` in scheduler tests. If needed, I can create a helper function to reduce the complexity of those repeated calls. That should allow us to keep the related internals private.
For 2, I'd like to just expose the constructor by default rather than just for tests. It doesn't seem that dangerous to me.